### PR TITLE
Update GK usage

### LINKED ISSUE #295: Fix GK integration and move to TF
**What feature would you like to see added?**
We should move away from using zero-padded acfs and just use the TF implementation. This would require that we also have an integration range argument so that one can compute the acf over a longer range than they integrate. This is not only more correct but it means we can immediately use TF for the implementation and get that GPU performance we are looking for.

**Describe how this feature would improve the code**
It would be more scientifically correct and more flexible.

**Describe how this feature would work/possible implementations**
Change the acf implementation and include an integration_range argument.

### PR DESCRIPTION
**Pull request by SamTov**
Fix #295 

Moved the gk implementation to tensor flow so that it now runs in parallel and updated the interface.

#### Summary of additions and changes

* Updated the interface to allow for a custom integration range argument.
* Implemented auto-correlation in tensor flow
* Updated plotting to show the integration line with respect to the acf

#### Not addressed

* Correct implementation of the distinct coefficients. I still need to better understand how the simple TF correlation compares with scipy
* Correct test runs. Due to the introduction of pure acf computations we no longer have zero-padding. While this makes the code more accurate and flexible, we now will always required a lot of statistics to get converged results, even on fast systems like the molten salts. Therefore, we will need to update the data used in the tests to address this.

### CODE REVIEW

File: mdsuite/calculators/green_kubo_distinct_diffusion_coefficients.py [hunk 1]
```diff
@@ -152,15 +164,18 @@ def run_experimental_analysis(self):
...
             self.atom_selection = select_atoms
         for combination in self.combinations:
             type_spec = {}
             data_path = [join_path(item, 'Velocities') for item in combination]
             self._prepare_managers(data_path=data_path)
             type_spec = self._update_species_type_dict(type_spec, data_path, 3)
-            type_spec[str.encode('data_size')] = tf.TensorSpec(None, dtype=tf.int16)
-            batch_generator, batch_generator_args = self.data_manager.batch_generator(dictionary=True)
+            type_spec[str.encode('data_size')] = tf.TensorSpec(None,
+                                                               dtype=tf.int16)
```
Comment (PythonFZ) [inline mdsuite/calculators/green_kubo_distinct_diffusion_coefficients.py:176]: we should start using `self.dtype` - but here we are really going for int16? I don't thing int32 would be slower but safer longterm?

File: mdsuite/calculators/green_kubo_distinct_diffusion_coefficients.py [hunk 2]
```diff
@@ -231,27 +250,36 @@ def _post_operation_processes(self, species: Union[str, tuple] = None):
...
         self._update_properties_file(properties)
         # Update the plot if required
         if self.plot:
-            plt.plot(np.array(self.time) * self.experiment.units['time'], self.vacf, label=species)
+            plt.plot(np.array(self.time) * self.experiment.units['time'],
+                     self.vacf,
+                     label=species)
+            plt.vlines((np.array(self.time) * self.experiment.units['time'])[
+                           self.integration_range], min(self.vacf),
+                       max(self.vacf))
```
Comment (PythonFZ) [inline mdsuite/calculators/green_kubo_distinct_diffusion_coefficients.py:265]: I guess this will be changed to the visualization module eventually?

File: mdsuite/calculators/green_kubo_ionic_conductivity.py [hunk 3]
```diff
@@ -113,8 +123,10 @@ def _update_output_signatures(self):
         -------
 
         """
-        self.batch_output_signature = tf.TensorSpec(shape=(self.batch_size, 3), dtype=tf.float64)
-        self.ensemble_output_signature = tf.TensorSpec(shape=(self.data_range, 3), dtype=tf.float64)
+        self.batch_output_signature = tf.TensorSpec(shape=(self.batch_size, 3),
+                                                    dtype=tf.float64)
```
Comment (PythonFZ) [inline mdsuite/calculators/green_kubo_ionic_conductivity.py:127]: `self.dtype` should be a property that can be set to `dtype=tf.float32` or `float64` so we can change the accuracy of the calculators for performance reasons

File: mdsuite/calculators/green_kubo_ionic_conductivity.py [hunk 4]
```diff
@@ -155,11 +167,14 @@ def _apply_operation(self, ensemble, index):
...
-                                     method='auto') for idx in range(3)])
-        self.jacf += jacf[int(self.data_range - 1):]
-        self.sigma.append(np.trapz(jacf[int(self.data_range - 1):], x=self.time))
+        jacf = self.data_range * tf.reduce_sum(
+            tfp.stats.auto_correlation(ensemble,
+                                       normalize=False,
+                                       axis=0,
+                                       center=False), axis=-1)
+        self.jacf += jacf
+        self.sigma.append(np.trapz(jacf[:self.integration_range],
```
Comment (PythonFZ) [inline mdsuite/calculators/green_kubo_ionic_conductivity.py:176]: `tfp.math.trapz` could be used, so we can eventually move everything into graph mode with tf.function

File: mdsuite/calculators/green_kubo_viscosity.py [hunk 5]
```diff
@@ -119,13 +132,17 @@ def _calculate_prefactor(self, species: str = None):
...
-            self.experiment.units['time'] / self.experiment.units['energy']
+        denominator = 3 * (self.data_range - 1) * \
+                      self.experiment.temperature * \
+                      self.experiment.units['boltzman'] * \
+                      self.experiment.volume
+
+        prefactor_units = self.experiment.units['pressure'] ** 2 * \
+                          self.experiment.units['length'] ** 3 * \
+                          self.experiment.units['time'] / \
+                          self.experiment.units['energy']
```
Comment (PythonFZ) [inline mdsuite/calculators/green_kubo_viscosity.py:145]: we should make this a dataclass or a named tuple:

```py
from dataclasses import dataclass

@dataclass(frozen=True)
class Units:
    pressure: float = 1.23456
    length: float = 1.1234656
    
   def gk_prefactor(self):
       return self.pressure ** 2 * self.length ** 3 + ...
```

so we can have `self.experiment.units.pressure` or `self.experiment.units.gk_prefactor()`

**Review by PythonFZ — APPROVED** [review]
I have not tested it, but therefore we do have the tests running.

Looks all good, I have a small comment on especially the `self.experiment.units` dict, that we could consider changing, but that is not really related to this PR. If we agree on it, we should make it an issue.

File: mdsuite/calculators/green_kubo_distinct_diffusion_coefficients.py (hunk 1 above)
Comment (SamTov) [reply mdsuite/calculators/green_kubo_distinct_diffusion_coefficients.py:176]: absolutely. I will update that.

File: mdsuite/calculators/green_kubo_distinct_diffusion_coefficients.py (hunk 2 above)
Comment (SamTov) [reply mdsuite/calculators/green_kubo_distinct_diffusion_coefficients.py]: Yes. In fact the idea is to have a trigger in the plotting that allows for a re-calculation of the integral by user input.

File: mdsuite/calculators/green_kubo_ionic_conductivity.py (hunk 3 above)
Comment (SamTov) [reply mdsuite/calculators/green_kubo_ionic_conductivity.py]: Sure. When was this introduced? Because I saw the code when I was updating things but had no idea it was introduced. This is the kind of thing we need to have in the dev docs or something as without information about it it will be hard to always enforce.

File: mdsuite/calculators/green_kubo_viscosity.py (hunk 5 above)
Comment (SamTov) [reply mdsuite/calculators/green_kubo_viscosity.py]: Issue #229 might be a good place for this kind of discussion. I suggest there that we push globals to a data file but a class also sounds quite good. We should discuss the pros and cons. In general though this is a really good idea.

File: mdsuite/calculators/green_kubo_ionic_conductivity.py (hunk 3 above)
Comment (SamTov) [reply mdsuite/calculators/green_kubo_ionic_conductivity.py]: Let's open an issue to address this.

File: mdsuite/calculators/green_kubo_ionic_conductivity.py (hunk 4 above)
Comment (SamTov) [reply mdsuite/calculators/green_kubo_ionic_conductivity.py]: This is only possible in the newest version of tfp which is not yet full deployed for all architectures.
